activerecord/lib/active_record/tasks/sqlite_database_tasks.rb

```diff
@@ -54,36 +46,32 @@ def structure_dump(filename, extra_flags)
         else
           args << ".schema --nosys"
         end
-        run_cmd("sqlite3", args, filename)
+
+        run_cmd("sqlite3", *args, out: filename)
       end
 
       def structure_load(filename, extra_flags)
         flags = extra_flags.join(" ") if extra_flags
         `sqlite3 #{flags} #{db_config.database} < "#{filename}"`
       end
 
-      private
-        attr_reader :db_config, :root
-
-        def connection
-          ActiveRecord::Base.lease_connection
+      def check_current_protected_environment!(db_config, migration_class)
+        super
+      rescue ActiveRecord::StatementInvalid => e
+        case e.message
+        when /SQLite3::ReadOnlyException/
```

Comment:
Have you try matching `e.cause` instead? Regexp on error messages isn't the end of the world, but is best avoided.